### PR TITLE
[CELEBORN-776] Restore package name of MasterNotLeaderException

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

Restore the package name of `MasterNotLeaderException`.

- 0.2
```
org.apache.celeborn.common.haclient.MasterNotLeaderException
```

- 0.3
```
org.apache.celeborn.common.client.MasterNotLeaderException
org.apache.celeborn.common.haclient.MasterNotLeaderException (server return this one)
```

- 0.4
```
org.apache.celeborn.common.client.MasterNotLeaderException (server return this one)
```

### Why are the changes needed?

```
2023-07-06 16:12:24 [WARN] [rpc-client-1-1] org.apache.celeborn.common.network.server.TransportChannelHandler#77 - Exception in connection from celeborn-master-0.celeborn-master-svc.spark/10.153.153.23:9097
java.lang.ClassNotFoundException: org.apache.celeborn.common.client.MasterNotLeaderException
	at java.net.URLClassLoader.findClass(URLClassLoader.java:387) ~[?:1.8.0_352]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_352]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_352]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_352]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_352]
	at org.apache.celeborn.common.serializer.JavaDeserializationStream$$anon$1.resolveClass(JavaSerializer.scala:68) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1988) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1852) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2186) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1669) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2431) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2355) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2213) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1669) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:503) ~[?:1.8.0_352]
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:461) ~[?:1.8.0_352]
	at org.apache.celeborn.common.serializer.JavaDeserializationStream.readObject(JavaSerializer.scala:76) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.serializer.JavaSerializerInstance.deserialize(JavaSerializer.scala:110) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.rpc.netty.NettyRpcEnv.$anonfun$deserialize$2(NettyRpcEnv.scala:276) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62) ~[scala-library-2.12.15.jar:?]
	at org.apache.celeborn.common.rpc.netty.NettyRpcEnv.deserialize(NettyRpcEnv.scala:323) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.rpc.netty.NettyRpcEnv.$anonfun$deserialize$1(NettyRpcEnv.scala:275) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62) ~[scala-library-2.12.15.jar:?]
	at org.apache.celeborn.common.rpc.netty.NettyRpcEnv.deserialize(NettyRpcEnv.scala:275) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.rpc.netty.NettyRpcEnv.$anonfun$ask$6(NettyRpcEnv.scala:235) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.rpc.netty.NettyRpcEnv.$anonfun$ask$6$adapted(NettyRpcEnv.scala:235) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.rpc.netty.RpcOutboxMessage.onSuccess(Outbox.scala:82) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.network.client.TransportResponseHandler.handle(TransportResponseHandler.java:180) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.network.server.TransportChannelHandler.channelRead(TransportChannelHandler.java:119) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.common.network.util.TransportFrameDecoder.channelRead(TransportFrameDecoder.java:74) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at org.apache.celeborn.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[celeborn-client-spark-3-shaded_2.12-0.2.0.13.jar:?]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_352]
```

### Does this PR introduce _any_ user-facing change?

Yes, bring back compatibility with 0.2.1 client

### How was this patch tested?

Use 0.2.1 client with branch-0.3 master/worker

### CODE REVIEW

**Comment by pan3793** [comment]
cc @cfmcgrady

File: common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.common.haclient;
+
+import java.io.IOException;
+
+// This class is reserved for compatible with 0.2 client
+// TODO remove it in 0.4.0
```
Comment (cfmcgrady) [inline common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java:23]: Does this mean the PR target is branch-0.3?

Comment (pan3793) [reply common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java:23]: Yes, this should be included in 0.3 to gain compatibility with 0.2 clients.

Comment (cfmcgrady) [reply common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java:23]: Then the 0.4 server does not gain compatibility with 0.2 clients? What compatibility strategy should we provide?

Comment (pan3793) [reply common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java:23]: Before 1.0
- 0.2 client support 0.2 and 0.3 server
- 0.3 client support 0.3 and 0.4 server
- ...

Comment (cfmcgrady) [reply common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java:23]: thank you for clarifying.

then change the PR target branch from `main` to `branch-0.3`?

**Review by cfmcgrady — APPROVED** [review]

**Review by AngersZhuuuu — APPROVED** [review]
LGTM, we should add a TODO for removing this after 0.3.0

**Comment by pan3793** [comment]
As suggested by @cfmcgrady, we can apply this patch only for `branch-0.3`

File: common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java (hunk 1 above)
Comment (pan3793) [reply common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java]: changed

Comment (FMX) [reply common/src/main/java/org/apache/celeborn/common/haclient/MasterNotLeaderException.java]: > Before 1.0
> 
> * 0.2 client support 0.2 and 0.3 server
> * 0.3 client support 0.3 and 0.4 server
> * ...

+1

**Comment by pan3793** [comment]
Closed via https://github.com/apache/incubator-celeborn/commit/0703980842567b48bcea98aaa0c5de07d9b72290

**Comment by zhongqiangczq** [comment]
> cc @zhongqiangczq, this is an issue, it's expected to be included in 0.3.0

ok

**Pull request closed without merging**